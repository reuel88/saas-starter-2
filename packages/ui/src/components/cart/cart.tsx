import { FC } from "react";
import { Button, buttonVariants } from "../button";
import {
  Sheet,
  SheetContent,
  SheetFooter,
  SheetHeader,
  SheetTrigger,
} from "../ui/sheet";
import { ShoppingCart, X } from "lucide-react";

type CartProps = {
  cart: Partial<{
    id: number;
    products: {
      id: number;
      title: string;
      price: number;
      quantity: number;
      total: number;
      discountPercentage: number;
      discountedPrice: number;
      thumbnail: string;
    }[];
    total: number;
    discountedTotal: number;
    userId: number;
    totalProducts: number;
    totalQuantity: number;
  }>;
};

export const Cart: FC<CartProps> = ({ cart }) => {
  const { products, total } = cart;

  return (
    <>
      <Sheet>
        <SheetTrigger
          className={buttonVariants({ variant: "outline" })}
          data-testid="checkout"
        >
          <ShoppingCart className="h-5 w-5" />
          <X className="h-5 w-5 text-muted-foreground" />
          <span className="sr-only">Checkout</span>
          <span className="text-muted-foreground">{products?.length ?? 0}</span>
          <span className="sr-only">items</span>
        </SheetTrigger>
        <SheetContent>
          <SheetHeader>
            <h2 className="text-lg font-medium">Shopping Cart</h2>
          </SheetHeader>
          <div>
            {products?.map((product) => (
              <article
                key={product.id}
                className="flex items-center space-x-4 border-b py-6"
              >
                <img
                  src={product.thumbnail}
                  alt={product.title}
                  className="h-16 w-16 rounded-md object-cover"
                />
                <div className="flex-1">
                  <h3 className="line-clamp-1 font-medium text-muted-foreground">
                    {product.title}
                  </h3>

                  <div className="flex items-end justify-between">
                    <div>
                      <p className="text-sm text-muted-foreground">
                        {product.quantity} x{" "}
                        {new Intl.NumberFormat("en-AU", {
                          style: "currency",
                          currency: "AUD",
                          minimumFractionDigits: 0,
                        }).format(product.price)}
                      </p>
                      <strong>
                        {new Intl.NumberFormat("en-AU", {
                          style: "currency",
                          currency: "AUD",
                          minimumFractionDigits: 0,
                        }).format(product.total)}
                      </strong>
                    </div>

                    <button
                      type="button"
                      className="text-muted-foreground hover:underline "
                    >
                      Remove
                    </button>
                  </div>
                </div>
              </article>
            ))}
          </div>
          <SheetFooter className="flex-col items-center space-y-0 sm:flex-col sm:justify-normal sm:space-x-0">
            <strong className="py-4 text-lg font-medium">
              Order Total:{" "}
              {new Intl.NumberFormat("en-AU", {
                style: "currency",
                currency: "AUD",
                minimumFractionDigits: 0,
              }).format(total ?? 0)}
            </strong>
            <Button className="w-full">Checkout</Button>
          </SheetFooter>
        </SheetContent>
      </Sheet>
    </>
  );
};
