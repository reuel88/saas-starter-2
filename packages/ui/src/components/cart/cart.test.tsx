import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Cart } from "./cart";

describe("Container", () => {
  it("should render a button with a shopping cart icon and the number of items in the cart", () => {
    // Arrange
    const cart = {
      products: [
        {
          id: 1,
          title: "Product 1",
          price: 10,
          quantity: 2,
          total: 20,
          discountPercentage: 0,
          discountedPrice: 0,
          thumbnail: "product1.jpg",
        },
        {
          id: 2,
          title: "Product 2",
          price: 15,
          quantity: 3,
          total: 45,
          discountPercentage: 0,
          discountedPrice: 0,
          thumbnail: "product2.jpg",
        },
      ],
      total: 65,
    };

    // Act
    const { baseElement } = render(<Cart cart={cart} />);

    // Assert
    expect(screen.getByText("Checkout")).toBeInTheDocument();
    expect(screen.getByText("2")).toBeInTheDocument();

    expect(baseElement).toMatchSnapshot();
  });

  it("should display a sheet when the button is clicked", async () => {
    // Arrange
    const cart = {
      products: [
        {
          id: 1,
          title: "Product 1",
          price: 10,
          quantity: 2,
          total: 20,
          discountPercentage: 0,
          discountedPrice: 0,
          thumbnail: "product1.jpg",
        },
      ],
      total: 20,
    };

    // Act
    const { baseElement } = render(<Cart cart={cart} />);
    const checkoutButton = screen.getByTestId("checkout");
    await userEvent.click(checkoutButton);

    // Assert
    expect(screen.getByText("Shopping Cart")).toBeInTheDocument();

    expect(baseElement).toMatchSnapshot();
  });
});
