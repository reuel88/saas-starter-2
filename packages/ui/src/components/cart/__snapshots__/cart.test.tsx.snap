// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Container > should display a sheet when the button is clicked 1`] = `
<body
  data-scroll-locked=""
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <button
      aria-controls="radix-:r3:"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
      data-state="open"
      data-testid="checkout"
      type="button"
    >
      <svg
        class="lucide lucide-shopping-cart h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="8"
          cy="21"
          r="1"
        />
        <circle
          cx="19"
          cy="21"
          r="1"
        />
        <path
          d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
        />
      </svg>
      <svg
        class="lucide lucide-x h-5 w-5 text-muted-foreground"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6 6 18"
        />
        <path
          d="m6 6 12 12"
        />
      </svg>
      <span
        class="sr-only"
      >
        Checkout
      </span>
      <span
        class="text-muted-foreground"
      >
        1
      </span>
      <span
        class="sr-only"
      >
        items
      </span>
    </button>
  </div>
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/80 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
  />
  <div
    aria-describedby="radix-:r5:"
    aria-labelledby="radix-:r4:"
    class="fixed z-50 gap-4 bg-background p-6 shadow-lg transition ease-in-out data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:duration-300 data-[state=open]:duration-500 inset-y-0 right-0 h-full w-3/4 border-l data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-sm"
    data-state="open"
    id="radix-:r3:"
    role="dialog"
    style="pointer-events: auto;"
    tabindex="-1"
  >
    <div
      class="flex flex-col space-y-2 text-center sm:text-left"
    >
      <h2
        class="text-lg font-medium"
      >
        Shopping Cart
      </h2>
    </div>
    <div>
      <article
        class="flex items-center space-x-4 border-b py-6"
      >
        <img
          alt="Product 1"
          class="h-16 w-16 rounded-md object-cover"
          src="product1.jpg"
        />
        <div
          class="flex-1"
        >
          <h3
            class="line-clamp-1 font-medium text-muted-foreground"
          >
            Product 1
          </h3>
          <div
            class="flex items-end justify-between"
          >
            <div>
              <p
                class="text-sm text-muted-foreground"
              >
                2
                 x
                 
                $10
              </p>
              <strong>
                $20
              </strong>
            </div>
            <button
              class="text-muted-foreground hover:underline "
              type="button"
            >
              Remove
            </button>
          </div>
        </div>
      </article>
    </div>
    <div
      class="flex flex-col items-center space-y-0 sm:flex-col sm:justify-normal sm:space-x-0"
    >
      <strong
        class="py-4 text-lg font-medium"
      >
        Order Total:
         
        $20
      </strong>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
      >
        Checkout
      </button>
    </div>
    <button
      class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-secondary"
      type="button"
    >
      <svg
        class="lucide lucide-x h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6 6 18"
        />
        <path
          d="m6 6 12 12"
        />
      </svg>
      <span
        class="sr-only"
      >
        Close
      </span>
    </button>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`Container > should render a button with a shopping cart icon and the number of items in the cart 1`] = `
<body>
  <div>
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
      data-state="closed"
      data-testid="checkout"
      type="button"
    >
      <svg
        class="lucide lucide-shopping-cart h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="8"
          cy="21"
          r="1"
        />
        <circle
          cx="19"
          cy="21"
          r="1"
        />
        <path
          d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
        />
      </svg>
      <svg
        class="lucide lucide-x h-5 w-5 text-muted-foreground"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6 6 18"
        />
        <path
          d="m6 6 12 12"
        />
      </svg>
      <span
        class="sr-only"
      >
        Checkout
      </span>
      <span
        class="text-muted-foreground"
      >
        2
      </span>
      <span
        class="sr-only"
      >
        items
      </span>
    </button>
  </div>
</body>
`;
